### PR TITLE
Add location field-type

### PR DESCRIPTION
**Pull request by nnatter**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Related issues/PRs | https://github.com/sulu/skeleton/pull/8
| License | MIT

#### What's in this PR?

This PR adds a react implementation of the `location` field-type.

#### Why?

Because we had a similar content type in Sulu 1.6.

#### Example Usage

~~~xml
<property name="location" type="location">
    <meta>
        <title lang="en">Location</title>
        <title lang="de">Ort</title>
    </meta>
</property>
~~~

### CODE REVIEW

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/Location.js [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+                                    <div>{value.street} {value.number}</div>
+                                    <div>{value.code} {value.town}</div>
+                                    <div>{value.country}</div>
+                                </Tooltip>
+                            }
+                        </Marker>
+                    </Map>
+                }
+                <LocationOverlay
+                    initialValue={value}
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/Location.js]: Why not just value and use it even if the value changes afterwards?

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js [hunk 2]
```diff
@@ -0,0 +1,261 @@
+// @flow
+import React from 'react';
+import {action, observable, reaction, toJS} from 'mobx';
+import {observer} from 'mobx-react';
+import {Form, Input, Number} from 'sulu-admin-bundle/components';
+import Overlay from 'sulu-admin-bundle/components/Overlay';
+import {translate} from 'sulu-admin-bundle/utils/Translator';
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: Check imports in this file :slightly_smiling_face:

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    @observable markerLong: ?number;
+
+    @observable title: ?string;
+    @observable street: ?string;
+    @observable number: ?string;
+    @observable code: ?string;
+    @observable town: ?string;
+    @observable country: ?string;
+
+    updateInitialDataDisposer: () => *;
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: use `() => void` (nothing happens with the return value, right?)

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+                this.mapZoom = this.props.initialValue ? this.props.initialValue.zoom : 1;
+
+                this.markerLat = this.props.initialValue ? this.props.initialValue.lat : null;
+                this.markerLong = this.props.initialValue ? this.props.initialValue.long : null;
+
+                this.title = this.props.initialValue ? this.props.initialValue.title : null;
+                this.street = this.props.initialValue ? this.props.initialValue.street : null;
+                this.number = this.props.initialValue ? this.props.initialValue.number : null;
+                this.code = this.props.initialValue ? this.props.initialValue.code : null;
+                this.town = this.props.initialValue ? this.props.initialValue.town : null;
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: Seems like you are already copying over the values anyway, so it's not really only the `initialValue`.

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+    componentWillUnmount() {
+        this.updateInitialDataDisposer();
+    }
+
+    handleConfirm = () => {
+        const {title, street, number, code, town, country, markerLat, markerLong, mapZoom} = this;
+
+        if (!markerLat || !markerLong) {
+            this.props.onConfirm(null);
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: Use destructuring before

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+            town,
+            country,
+            lat: markerLat,
+            long: markerLong,
+            zoom: mapZoom,
+        });
+    };
+
+    @action handleAutoCompleteChange = (data: Object) => {
+        console.log('select', toJS(data));
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: No `console.log` :) I hope that is also covered by eslint.

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    };
+
+    render() {
+        const {
+            onClose,
+            open,
+        } = this.props;
+
+        // enable confirm button if all marker properties are set or no property is set in case of a reset
+        const confirmEnabled = (this.markerLat && this.markerLong) || (!this.markerLat && !this.markerLong);
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: Not sure about this, haven't we handled this in other overlays in a different way? I would say we should only enable the confirm button if something has changed.

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/location.scss [hunk 8]
```diff
@@ -0,0 +1,55 @@
+@import 'sulu-admin-bundle/containers/Application/colors.scss';
+
+$mapBorderColor: $alto;
+$labelTextColor: $scorpion;
+$contentBackgroundColor: $alto;
+$buttonTextColor: $scorpion;
+$buttonTextColorActive: $white;
+$buttonBackgroundColor: $alto;
+$buttonBackgroundColorActive: $shakespeare;
+$buttonMargin: 2px;
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/location.scss]: Prefix the variables with `location`

File: src/Sulu/Bundle/LocationBundle/Resources/js/index.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
+// @flow
+import {fieldRegistry} from 'sulu-admin-bundle/containers';
+import L from 'leaflet';
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/index.js]: Isn't a one-letter import weird?

File: src/Sulu/Bundle/LocationBundle/Resources/js/index.js [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import leafletMarkerIcon from 'leaflet/dist/images/marker-icon.png';
+import leafletMarkerShadow from 'leaflet/dist/images/marker-shadow.png';
+import {Location} from './containers/Form';
+
+// leaflet requires that its stylesheet is embedded somewhere into the build
+// eslint-disable-next-line
```
Comment (danrot) [inline src/Sulu/Bundle/LocationBundle/Resources/js/index.js]: What rule is disabled here? Add the rule name after the comment, then the other rules still apply.

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js (hunk 3 above)
Comment (nnatter) [reply src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: unfortunately mobx declares the return type of `reaction` as `() => mixed`, therefore i used the same type as in other parts of the sulu code 🙂  https://github.com/mobxjs/mobx/blob/mobx4-master/flow-typed/mobx.js#L326

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js (hunk 4 above)
Comment (nnatter) [reply src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: would you call this just `value` then?

File: src/Sulu/Bundle/LocationBundle/Resources/js/index.js (hunk 9 above)
Comment (nnatter) [reply src/Sulu/Bundle/LocationBundle/Resources/js/index.js]: absolutely 😂 but leaflet uses this style throughout its documentation. we can still change it, dont have any opinion here. https://leafletjs.com/examples/quick-start/

File: src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js (hunk 7 above)
Comment (nnatter) [reply src/Sulu/Bundle/LocationBundle/Resources/js/containers/Location/LocationOverlay.js]: i would like to enable the confirm button only if the overlay contains consistent/valid values. we use something similar in the `ExternalLinkOverlay` o the `FocusPointOverlay`. but i could add an additional `dirty` observable/computed if you want to check that too 🙂

**Pull request merged**